Add support for window controls (aka decorationLayout) to headerbar

PhilippMDoerner opened the pull request.

can-lehmann commented:
From what I can tell, one of the major applications of `WindowControls` is creating split header bars (like we had in the `Slider` example at some point). The problem they solve here is allowing the user to select on which side of the window the close button should be shown. This is also why the `side` field exists.

For this application we need to provide the option to not override [GtkSettings:gtk-decoration-layout](https://gnome.pages.gitlab.gnome.org/gtk/gtk4/property.Settings.gtk-decoration-layout.html).

PhilippMDoerner commented:
Could you expand what conclusion you want me to draw from this ? Mostly because I can draw multiple.

E.g. the simplest interpretation would be to say "if seq[WindowControlButton].len == 0 then either don't set decoration-layout string or set it to the default-value" (both achieve the same result of having the default value present).

I'm just not sure if you mean that or want me to go in the direction of splitting "WindowControlButton" into "Left" and "right" or want me to abandon the approach with a "WindowControlButton-enum" altogether.

can-lehmann commented:
> Could you expand what conclusion you want me to draw from this ? Mostly because I can draw multiple.

The main point is allowing the user to explicitly leave/make unset. Note that this unset state is explicity different from adding no buttons. One way to do this is using `buttons: Option[seq[WindowControlButton]]` (or as below `buttons: Option[DecorationLayout]`). It needs to explicitly read `gtk-decoration-layout` if left unset.

Other related changes that might be interesting:
1. Create a "`DecorationLayout`" type that has fields for `left`/`right`
2. Make `buttons` a setter is overloaded to support both `DecorationLayout` (and immediately calls `$` on it) and a regular `string`

PhilippMDoerner commented:
I have just now learned that Headerbar **also** has a decorationLayout string.
Doing effectively the exact same.

Could it make sense to abandon wrapping `WindowControls` for the same reason that wrapping `Bin` was abandoned (it's kinda pointless) and instead add wrapping for decorationLayout to `HeaderBar`?

Because the way I currently see it nobody should be using the WindowControls widget if they can just use HeaderBar

PhilippMDoerner commented:
I've updated the PR to reflect what we just discussed in discord:

This PR is no longer for adding the windowcontrols widget.
It is for adding support of those window controls (including a nicer setter) to Headerbar.
I removed the example since headerbar previously didn't have one so I assume it was deemed too small to be worth an example widget.

can-lehmann commented:
I just saw that [Gtk.HeaderBar:decoration-layout](https://docs.gtk.org/gtk4/property.HeaderBar.decoration-layout.html) also overrides [GtkSettings:gtk-decoration-layout](https://docs.gtk.org/gtk4/property.Settings.gtk-decoration-layout.html). So the field should by of type `Option[string]` to allow for explicitly not overriding the property.

PhilippMDoerner commented on owlkettle/widgets.nim:
```diff
@@ -979,12 +979,29 @@ proc updateChild*(state: Renderable,
   left: seq[Widget]
   right: seq[Widget]
   
+  setter windowControls: DecorationLayout
```
I don't get that one. If you don't want to set a decorationlayout, just don't use the setter no?

PhilippMDoerner replied:
I coded it in despite not really getting why.
Note that a specific variation of the syntax **will** cause completely garbage compiler errors.

Basically, if any of the seq's in the tuple are empty and you instantiate it as a `some` optional in a single line, it'll always cause this error:
`Error: unhandled exception: liftdestructors.nim(602, 5) `t.destructor != nil`  [AssertionDefect]`

Examples where this blows up if you add this e.g. to the scale.nim example:
```nim
#examples are before the gui block, but you'll get the same if you do it inside of it as windowControls = some etc.
let controls: Option[DecorationLayout] = some((@[WindowControlIcon], @[]))
let controls: Option[DecorationLayout] = some((@[], @[]))
let controls: Option[DecorationLayout] = some((@[], @[WindowControlIcon]))
```

PhilippMDoerner commented:
@can-lehmann I'm closing this PR and merging it into #116 .
That PR with AdwHeaderBar also needs decorationlayout and basically all the same hooks, enums and toLayoutString procs etc.

It makes sense to not develop double / have to refactor that after merge and instead combine them right now.
As per your previous statement that you prefer "coupled" PRs to be combined, I am thus combining them.
